### ISSUE TITLE
[rllib] Setting `training_enabled` in `external_env._ExternalEnvEpisode()` does not do anything

### ISSUE THREAD

**Issue by heng2j** [ray-project/ray#17463]
### What is the problem?

Setting `training_enabled` in `external_env._ExternalEnvEpisode()` does not do anything but simply updating the info dict.

This issue had been documented in Ray discuss and verified by @sven1977 

https://discuss.ray.io/t/rllib-how-does-setting-training-enabled-to-false-in-external-env-externalenvepisode-trigger-no-training/3059

*Ray version and other system information (Python version, TensorFlow version, OS):*

Ubuntu 18.04 LTS
Ray 1.3.0
Python 3.7.5
tensorflow-gpu 2.3.1

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

This issue can't be reproduced with less than 50 lines of code as it required the RLlib Policy client server pattern.

To reproduce the issue:

Run the [cartpole_server](https://github.com/ray-project/ray/blob/master/rllib/examples/serving/cartpole_server.py) and  [cartpole_client](https://github.com/ray-project/ray/blob/master/rllib/examples/serving/cartpole_client.py) example twice one setting `training_enabled` in client to `False` and one set to `True`. And but same random seed, same config and same training iteration. And you will see identical results which is not ideal.

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by heinerb** [comment]
Hello @sven1977 Question on this item and possible fix for the issue. 

**Original Conversation**:  https://discuss.ray.io/t/rllib-how-does-setting-training-enabled-to-false-in-external-env-externalenvepisode-trigger-no-training/3059

**Suggested update - [external_env.py](https://github.com/ray-project/ray/blob/ddad6a2458bded21035797ec84ba67b8a8c15ca5/rllib/env/external_env.py#L252)** 

The following code update would allow setting training enabled info flag at each step in the information dict.
  
```python
    def _send(self):
        if self.multiagent:
            if self.cur_info_dict:
                for agent_id in self.cur_info_dict:
                    self.cur_info_dict[agent_id]["training_enabled"] = self.training_enabled
            else:
                for agent_id in self.new_observation_dict:
                    self.cur_info_dict[agent_id] = {}
                    self.cur_info_dict[agent_id]["training_enabled"] = self.training_enabled
            item = {
                "obs": self.new_observation_dict,
                "reward": self.cur_reward_dict,
                "done": self.cur_done_dict,
                "info": self.cur_info_dict,
            }
            if self.new_action_dict is not None:
                item["off_policy_action"] = self.new_action_dict
            self.new_observation_dict = None
            self.new_action_dict = None
            self.cur_reward_dict = {}
        else:
            item = {
                "obs": self.new_observation,
                "reward": self.cur_reward,
                "done": self.cur_done,
                "info": self.cur_info,
            }
            if self.new_action is not None:
                item["off_policy_action"] = self.new_action
            self.new_observation = None
            self.new_action = None
            self.cur_reward = 0.0
            item["info"]["training_enabled"] = self.training_enabled

        with self.results_avail_condition:
            self.data_queue.put_nowait(item)
            self.results_avail_condition.notify()
```

With every step now having the flag set in information dict the [sampler.py:1029](https://github.com/ray-project/ray/blob/ddad6a2458bded21035797ec84ba67b8a8c15ca5/rllib/evaluation/sampler.py#L1031) will have access to access to the flag in the eval_data before calling `compute_actions_from_input_dict`. The flag could be passed to `explore` argument allowing disabling explore when set. 

```python
eval_results[policy_id] = \
            policy.compute_actions_from_input_dict(
                input_dict,
                timestep=policy.global_timestep,
                episodes=[active_episodes[t.env_id] for t in eval_data],
                explore=eval_data[0].info.get("training_enabled", True))
```

**Comment by antoine-galataud** [comment]
@heng2j @sven1977 I've just discovered this on a policy server I've upgraded from 0.8.6 to 1.4.0. This server isn't supposed to train continuously but does (`training_enabled = False` when client calls the server).
I think the fix must happen in `sampler.py`, by conditioning collection of a new sample with a check on this flag.
